Sources/Conduit/Auth/Cryptography/Symmetric/KeychainHybridKeyProvider.swift

```diff
@@ -0,0 +1,186 @@
+//
+//  KeychainHybridKeyProvider.swift
+//  Conduit
+//
+//  Created by John Hammerlund on 12/10/19.
+//
+
+import Foundation
+import Security
+#if os(iOS) || os(macOS)
+import LocalAuthentication
+#endif
+
+/// Provides contextual hybrid-encryption keys via Keychain queries. On supported devices, elliptic-curve keys are stored
+/// on the Secure Enclave, providing the maximum application security for key management.
+@available(macOS 10.12.1, iOS 10.0, tvOS 10.0, watchOS 3.0, *)
+public final class KeychainHybridKeyProvider: HybridKeyProvider {
+
+    private static let itemPrefix = "com.mindbodyonline.Conduit.hybrid-key.secret"
+    private static let privateItemLabel = "com.mindbodyonline.Conduit.hybrid-key.private"
+    private static let publicItemLabel = "com.mindbodyonline.Conduit.hybrid-key.public"
+
+    private let accessGroup: String?
+    private let accessibility: CFString
+    private let itemTag: String
+    private lazy var tagData: Data? = {
+        itemTag.data(using: .utf8)
+    }()
+    private let encryptionType: HybridEncryptionType
+    public private(set) lazy var keyAlgorithm: SecKeyAlgorithm = encryptionType.algorithm
+    /// If true, then applicable keys will be stored on the Secure Enclave on supported devices. Defaults to `true`
+    /// - Important: Mac applications require entitlements (and therefore, valid codesigning) in order to support this.
+    ///   Otherwise, this will trigger an `errSecMissingEntitlements`.
+    public var prefersSecureEnclaveStorage: Bool = true
+
+    private lazy var itemQuery: [String: Any] = {
+        var query: [String: Any] = [
+            kSecClass as String: kSecClassKey,
+            kSecAttrKeyType as String: encryptionType.keyType
+        ]
+        query[kSecAttrAccessGroup as String] = accessGroup
+        query[kSecAttrApplicationTag as String] = tagData
+        return query
+    }()
+    private lazy var canUseSecureEnclave: Bool = {
+        guard prefersSecureEnclaveStorage else {
+            return false
+        }
+        #if !targetEnvironment(simulator) && !os(watchOS) && !os(tvOS)
+        if encryptionType.supportsSecureEnclaveStorage, #available(OSX 10.12.2, *) {
+            // Unfortunately, the only way to determine if the secure enclave is available is via determining biometric capabilities.
+            // Attempting to store items in the secure enclave on unsupported devices will result in a crash on certain versions.
+            var error: NSError?
+            if LAContext().canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) {
+                return true
+            }
+            guard let errorCode = error.flatMap({ LAError.Code(rawValue: $0.code) }) else {
+                return false
+            }
+            if #available(OSX 10.13, iOS 11, *) {
```

Comment:
How does this code run on iOS, when OS X 10.12.2 is required above? ` #available(OSX 10.12.2, *)`

Reply:
`if #available(OSX 10.12.2, *)` checks if you're running on macOS 10.12.2 or above OR on any other platform via the `*`